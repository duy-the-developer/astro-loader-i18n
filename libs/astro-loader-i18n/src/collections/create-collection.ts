import { i18nLoaderSchema } from "../schemas/i18n-loader-schema";
import { z } from "astro/zod";

type Options = {
  locales: string[];
  routePattern: string;
};

export function createCollection(options: Options): z.infer<typeof i18nLoaderSchema>[] {
  const { locales, routePattern } = options;
  return locales.map((locale) => ({ locale: locale, translationId: routePattern }));
}
