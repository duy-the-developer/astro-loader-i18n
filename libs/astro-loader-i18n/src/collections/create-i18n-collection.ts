import { i18nLoaderSchema } from "../schemas/i18n-loader-schema";
import { z } from "astro/zod";

type Options = {
  locales: string[];
  routePattern: string;
};

export type I18nCollection = { data: z.infer<typeof i18nLoaderSchema> }[];

export function createI18nCollection(options: Options): I18nCollection {
  const { locales, routePattern } = options;

  return locales.map((locale) => ({ data: { locale: locale, translationId: routePattern, contentPath: "" } }));
}
