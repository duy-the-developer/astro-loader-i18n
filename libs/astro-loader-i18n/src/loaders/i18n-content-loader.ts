import { glob, Loader, LoaderContext } from "astro/loaders";
import { pruneLocales, getAllUniqueKeys } from "../utils/collection";
import { createTranslationId } from "../utils/path";

type GlobOptions = Parameters<typeof glob>[0];

export function i18nContentLoader(options: GlobOptions): Loader {
  const globLoader = glob(options);
  return {
    name: "i18n-content-loader",
    load: async (context: LoaderContext) => {
      if (!context.config.i18n) throw new Error("i18n configuration is missing in your astro config");

      const { locales, defaultLocale } = context.config.i18n;
      const localeCodes = locales.flatMap((locale) => (typeof locale === "string" ? locale : locale.codes));

      const parseData = context.parseData;
      const parseDataProxy: typeof parseData = (props) => {
        if (!props.filePath) return parseData(props);
        const locale = defaultLocale;
        const translationId = createTranslationId(props.filePath);
        return parseData({ ...props, data: { ...props.data, locale, translationId } });
      };
      context.parseData = parseDataProxy;

      await globLoader.load(context);

      const entries = context.store.entries();
      context.store.clear();

      entries.forEach(([, entry]) => {
        const entryLocales = Array.from(getAllUniqueKeys(entry.data)).filter((key) => localeCodes.includes(key));
        if (entryLocales.length === 0) {
          context.store.set(entry);
        } else {
          entryLocales.forEach((locale) => {
            const entryData = pruneLocales(entry.data, entryLocales, locale);
            context.store.set({ ...entry, id: `${entry.id}/${locale}`, data: { ...entryData, locale } });
          });
        }
      });
    },
  };
}
