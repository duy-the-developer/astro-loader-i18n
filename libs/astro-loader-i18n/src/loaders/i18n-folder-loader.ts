import { glob, Loader, LoaderContext } from "astro/loaders";
import { createCommonTranslationId, parseLocale } from "../utils/path";

type GlobOptions = Parameters<typeof glob>[0];

export function i18nFolderLoader(options: GlobOptions): Loader {
  const globLoader = glob(options);
  return {
    name: "i18n-folder-loader",
    load: async (context: LoaderContext): Promise<void> => {
      await globLoader.load(context);
      const entries = context.store.values();

      if (!context.config.i18n) throw new Error("i18n configuration is missing in your astro config");

      const { locales, defaultLocale } = context.config.i18n;
      const localeCodes = locales.flatMap((locale) => (typeof locale === "string" ? locale : locale.codes));

      context.store.clear();
      entries.forEach((entry) => {
        if (!entry.filePath) return;
        const locale = parseLocale(entry.filePath, localeCodes, defaultLocale);

        const newEntry = {
          ...entry,
          data: {
            ...entry.data,
            locale,
            commonTranslationId: createCommonTranslationId(entry.filePath, locale),
          },
        };
        context.store.set(newEntry);
      });
    },
  };
}
