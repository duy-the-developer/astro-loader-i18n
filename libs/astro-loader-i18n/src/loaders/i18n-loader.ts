import { glob, Loader, LoaderContext } from "astro/loaders";
import { createTranslationId, parseLocale } from "../utils/path";

type GlobOptions = Parameters<typeof glob>[0];

export function i18nLoader(options: GlobOptions): Loader {
  const globLoader = glob(options);
  return {
    name: "i18n-loader",
    load: async (context: LoaderContext) => {
      if (!context.config.i18n) throw new Error("i18n configuration is missing in your astro config");

      const { locales, defaultLocale } = context.config.i18n;
      const localeCodes = locales.flatMap((locale) => (typeof locale === "string" ? locale : locale.codes));

      const parseDataProxy: typeof context.parseData = (props) => {
        if (!props.filePath) return context.parseData(props);
        const locale = parseLocale(props.filePath, localeCodes, defaultLocale);
        const translationId = createTranslationId(props.filePath, locale);
        return context.parseData({ ...props, data: { ...props.data, locale, translationId } });
      };

      const globContext = Object.assign({}, context, { parseData: parseDataProxy });
      await globLoader.load(globContext);
    },
  };
}
