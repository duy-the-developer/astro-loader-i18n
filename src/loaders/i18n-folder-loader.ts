import { glob, Loader, LoaderContext } from "astro/loaders";
import { GlobOptions } from "../types";

export function i18nFolderLoader(options: GlobOptions ): Loader {
  return {
    name: "i18n-folder-loader",
    load: async (context: LoaderContext): Promise<void> => {
      await glob(options).load(context);

    }
  }
}
