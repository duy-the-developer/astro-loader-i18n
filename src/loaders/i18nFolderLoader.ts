import { Loader, LoaderContext } from "astro/loaders";

export function i18nFolderLoader(): Loader {
  return {
    name: "i18n-folder-loader",
    load: async (context: LoaderContext): Promise<void> => {}
  }
}
