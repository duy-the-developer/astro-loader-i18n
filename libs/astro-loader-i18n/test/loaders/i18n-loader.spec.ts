import { beforeEach, describe, expect, it, vi } from "vitest";
import { i18nLoader } from "../../src/loaders/i18n-loader";
import { LoaderContext } from "astro/loaders";
import { createLoaderContext } from "../__mocks__/loader-context";
import { folderCollectionFixture } from "../__fixtures__/collections";

vi.mock("astro/loaders", () => {
  return {
    glob: () => {
      return {
        load: vi.fn().mockImplementation(async (context: LoaderContext) => {
          folderCollectionFixture.forEach(async (entry) => {
            context.store.set({ ...entry, data: await context.parseData(entry) });
          });
        }),
      };
    },
  };
});

describe("i18nLoader", () => {
  let context: LoaderContext;

  beforeEach(() => {
    context = createLoaderContext();
  });

  it("should put common translation id and locale in data", async () => {
    const loader = i18nLoader({ pattern: "**/*.mdx", base: "./src/content/pages" });
    await loader.load(context);

    const entries = context.store.entries();
    expect(entries).toMatchSnapshot();
  });

  it("should throw error if i18n config is missing", async () => {
    const loader = i18nLoader({ pattern: "**/*.mdx", base: "./src/content/pages" });
    context.config.i18n = undefined;

    await expect(loader.load(context)).rejects.toThrowErrorMatchingSnapshot();
  });

  it("should use locale.codes if locale is an object", async () => {
    const loader = i18nLoader({ pattern: "**/*.mdx", base: "./src/content/pages" });
    context.config.i18n = {
      defaultLocale: "zh-CN",
      routing: "manual",
      locales: [
        { codes: ["de-CH"], path: "/de" },
        { codes: ["en-US"], path: "/en" },
        { codes: ["zh-CN"], path: "/zh" },
      ],
    };
    await loader.load(context);

    const entries = context.store.entries();
    expect(entries).toMatchSnapshot();
  });
});
