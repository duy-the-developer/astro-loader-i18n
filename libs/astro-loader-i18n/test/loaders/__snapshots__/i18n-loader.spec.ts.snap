// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`i18nLoader > should put common translation id and locale in data 1`] = `
[
  [
    "de-CH/about",
    {
      "body": "---
title: Über mich
---

Test
",
      "data": {
        "basePath": "/",
        "contentPath": "",
        "locale": "de-CH",
        "title": "Über mich",
        "translationId": "src/content/pages/about.mdx",
      },
      "filePath": "src/content/pages/de-CH/about.mdx",
      "id": "de-CH/about",
    },
  ],
  [
    "de-CH/projects",
    {
      "body": "---
title: Projekte
---

Test
",
      "data": {
        "basePath": "/",
        "contentPath": "subpath",
        "locale": "de-CH",
        "title": "Projekte",
        "translationId": "src/content/pages/subpath/projects.mdx",
      },
      "filePath": "src/content/pages/de-CH/subpath/projects.mdx",
      "id": "de-CH/projects",
    },
  ],
  [
    "zh-CN/about",
    {
      "body": "---
title: 关于我
---

测试
",
      "data": {
        "basePath": "/",
        "contentPath": "",
        "locale": "zh-CN",
        "title": "关于我",
        "translationId": "src/content/pages/about.mdx",
      },
      "filePath": "src/content/pages/zh-CN/about.mdx",
      "id": "zh-CN/about",
    },
  ],
  [
    "zh-CN/projects",
    {
      "body": "---
title: 项目
---

测试
",
      "data": {
        "basePath": "/",
        "contentPath": "subpath",
        "locale": "zh-CN",
        "title": "项目",
        "translationId": "src/content/pages/subpath/projects.mdx",
      },
      "filePath": "src/content/pages/zh-CN/subpath/projects.mdx",
      "id": "zh-CN/projects",
    },
  ],
]
`;

exports[`i18nLoader > should throw error if i18n config is missing 1`] = `[Error: i18n configuration is missing in your astro config]`;

exports[`i18nLoader > should use locale.codes if locale is an object 1`] = `
[
  [
    "de-CH/about",
    {
      "body": "---
title: Über mich
---

Test
",
      "data": {
        "basePath": "/",
        "contentPath": "",
        "locale": "de-CH",
        "title": "Über mich",
        "translationId": "src/content/pages/about.mdx",
      },
      "filePath": "src/content/pages/de-CH/about.mdx",
      "id": "de-CH/about",
    },
  ],
  [
    "de-CH/projects",
    {
      "body": "---
title: Projekte
---

Test
",
      "data": {
        "basePath": "/",
        "contentPath": "subpath",
        "locale": "de-CH",
        "title": "Projekte",
        "translationId": "src/content/pages/subpath/projects.mdx",
      },
      "filePath": "src/content/pages/de-CH/subpath/projects.mdx",
      "id": "de-CH/projects",
    },
  ],
  [
    "zh-CN/about",
    {
      "body": "---
title: 关于我
---

测试
",
      "data": {
        "basePath": "/",
        "contentPath": "",
        "locale": "zh-CN",
        "title": "关于我",
        "translationId": "src/content/pages/about.mdx",
      },
      "filePath": "src/content/pages/zh-CN/about.mdx",
      "id": "zh-CN/about",
    },
  ],
  [
    "zh-CN/projects",
    {
      "body": "---
title: 项目
---

测试
",
      "data": {
        "basePath": "/",
        "contentPath": "subpath",
        "locale": "zh-CN",
        "title": "项目",
        "translationId": "src/content/pages/subpath/projects.mdx",
      },
      "filePath": "src/content/pages/zh-CN/subpath/projects.mdx",
      "id": "zh-CN/projects",
    },
  ],
]
`;
